ethexe/observer/src/query.rs

```diff
@@ -39,74 +45,207 @@ impl Query {
+            self.database.set_block_events(
+                block_hash,
+                blocks_events.remove(&block_hash).unwrap_or_default(),
+            );
+            committed_blocks.extend(self.get_committed_blocks(block_hash).await?);
+            block_hashes.push(block_hash);
+        }
+        block_hashes.reverse();
+        log::trace!("{} blocks loaded", block_hashes.len());
+
+        Ok((block_hashes, committed_blocks))
+    }
+
+    pub async fn load_chain(&mut self, from_hash: H256) -> Result<(Vec<H256>, Vec<H256>)> {
+        let mut chain = vec![];
+        let mut committed_blocks = Vec::new();
+        let mut hash = from_hash;
+        let mut header = self.get_block_header_meta(hash).await?;
 
+        loop {
+            // If the block's end state is valid, set it as the latest valid block
             if self
                 .database
                 .block_end_state_is_valid(hash)
                 .unwrap_or(false)
             {
+                self.database.set_latest_valid_block_height(header.height);
                 log::trace!("Nearest valid in db block found: {hash}");
                 break;
             }
 
             log::trace!("Include block {hash} in chain for processing");
+            committed_blocks.extend(self.get_committed_blocks(hash).await?);
             chain.push(hash);
 
-            committed_blocks.extend(self.get_committed_blocks(hash).await?);
+            match self.database.block_header(hash) {
+                Some(block_header) => {
+                    header = block_header;
+                    hash = header.parent_hash;
+                }
+                None => {
+                    hash = self.get_block_parent_hash(hash).await?;
+                }
+            }
+        }
 
-            hash = self.get_block_parent_hash(hash).await?;
+        Ok((chain, committed_blocks))
+    }
+
+    pub async fn get_last_committed_chain(&mut self, block_hash: H256) -> Result<Vec<H256>> {
+        let current_block = self.get_block_header_meta(block_hash).await?;
+        let latest_valid_block_height = self
+            .database
+            .latest_valid_block_height()
+            .expect("genesis by default; qed");
+
+        if current_block.height >= latest_valid_block_height
+            && (current_block.height - latest_valid_block_height) >= self.max_commitment_depth
+        {
+            return Err(anyhow!(
+                "Too deep chain: Current block height: {}, Latest valid block height: {}, Max depth: {}",
+                current_block.height,
+                latest_valid_block_height,
+                self.max_commitment_depth
+            ));
         }
 
+        // Determine if deep sync is needed
+        let is_deep_sync = {
+            // Current block can be lower than latest valid due to reorgs.
+            let block_diff = (current_block.height as i64 - latest_valid_block_height as i64)
+                .unsigned_abs() as u32;
```

Comment:
abs is useless. `saturating_sub` should be used. use it everywhere or `checked_sub`.